powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/ActivationServiceBehavior.java

```diff
@@ -1962,12 +1951,39 @@ private ActivationRecovery createRecoveryCodeForActivation(ActivationRecordEntit
         }
     }
 
+    /**
+     * Revoke recovery codes for an activation entity.
+     * @param activationId Activation identifier.
+     */
+    private void revokeRecoveryCodes(String activationId) {
+        logger.info("Revoking recovery codes for activation ID: {}", activationId);
+        final RecoveryCodeRepository recoveryCodeRepository = repositoryCatalogue.getRecoveryCodeRepository();
+        final List<RecoveryCodeEntity> recoveryCodeEntities = recoveryCodeRepository.findAllByActivationId(activationId);
+        final Date now = new Date();
+        for (RecoveryCodeEntity recoveryCode : recoveryCodeEntities) {
+            logger.debug("Revoking recovery code: {} for activation ID: {}", recoveryCode.getRecoveryCode(), activationId);
```

Comment:
Ok, I double checked that `RecoveryCodeEntity#toString` does not contain `recoveryCode` itself. So we may log the whole entity if the ID does not give enough context.